Chore/user level data

ivanThePleasant opened the pull request:
### What does it do?

This PR introduces refactored analytics that are now sending user-level context to the analytics server. Previously it was project-level context.

***Important***: users are still anonymous, analytics server only receives a hash of a user.

Changes that were introduced:
- All server and admin metrics calls now pass `adminUserId` property to their respective sender functions.
  - For admin metrics calls `adminUserId` is added at the sender function since admin user will be the same for the Admin Panel session
  - For server metrics calls, `adminUserId` is passed to every `strapi.telemetry.send()` call as the first argument. If the event supposed to be anonymous, empty string ( '' ) is passed in. 
- All generators events are by nature anonymous, but they still send an empty string as `adminUserId`
- Project uuid is now part of the `properties` object on all calls and has been renamed to `projectId`
- Tests were refactored to accommodate all these changes

### Why is it needed?

This is part of our initiative to migrate to user-level data analytics

### How to test it?

You can run the getstarted application and route all analytics events to your local server. You will be able to see the payload of the event.

***Apologies for a big PR, but there is pretty much no way to make it smaller as all these changes have to be merged at the same time.***

***Please don't merge this PR. I need it approved, but it will only be merged AFTER changes to the tracker are finalized and merged as well***

alexandrebodin commented on packages/core/admin/admin/src/utils/unique-admin-hash.js:
```diff
@@ -0,0 +1,7 @@
+const crypto = require('crypto');
```
crypto is a node package you don't want to use this in the frontend. Look into an equivalent with he browser APIs here

alexandrebodin commented on packages/plugins/i18n/server/services/metrics.js:
```diff
@@ -11,13 +11,13 @@ const sendDidInitializeEvent = async () => {
     0
```
those actions are made by users

ivanThePleasant replied:
They also fire every time the server starts, hence they are anonymous. Is that the desired behavior do you know?

alexandrebodin commented on packages/core/strapi/lib/services/metrics/index.js:
```diff
@@ -79,9 +80,9 @@ const createTelemetryInstance = (strapi) => {
       crons.forEach((cron) => cron.cancel());
     },
-    async send(event, payload) {
+    async send(adminUserId, event, payload) {
```
It would make the code a more readable to keep the previous format and accept adminUserId in the payload instead. That will also avoid any empty strange calls with `send('', 'event')`

ivanThePleasant replied:
I tried that, but then in the end I will have `adminUserId` twice on the payload to the tracker. It's going to be on the payload, which becomes `properties` later, so it will be on properties. And it also needs to be at the top of the body. You suggesting that I delete the key from `properties` object when it gets to the sender?

fdel-car replied:
I agree with Alex here.

Honestly, it's a good opportunity right now so I would go a bit further and rethink this interface and the tracker API, you can say no but as we're making those changes and most probably doing a v2 endpoint on the tracker or something similar it's a good window of opportunity.

This function prototype, I added typings for clarity:

```ts
interface Payload {
  userId?: string
  eventProperties?: Record<string, unknown>
  userProperties?: Record<string, unknown>
  groupProperties?: Record<string, unknown>
}

// Not sure we use this one anyway, we could remove it
interface Options {
  // fetch options here
}

function send(event: string, payload?: Payload, options?: Options);
```

With that interface we won't have to update the tracker everytime we want to add `groupProperties` or `userProperties`, more flexibility - and yes more responsponsability - for the API consumer, us.
And also we won't have the issues you mentioned above.

alexandrebodin replied:
@fdel-car's proposal looks pretty good.

alexandrebodin reviewed (changes requested):
A general comment I didn't know where to put:

- Can you please pass the user in the send telemetry payload instead of the hash so we only hash in the metris instead of everywhere in the code. This is the best way to keep some encapsulation

ivanThePleasant commented:
@alexandrebodin Thanks! So, just a quick recap, you'd prefer that hashing happens in the metrics service and events pass user with payload in order to hash it once there? Cool, let's do that 👍

ivanThePleasant commented on packages/core/strapi/lib/services/metrics/sender.js:
```diff
@@ -54,16 +54,17 @@ module.exports = (strapi) => {
       body: JSON.stringify({
         event,
-        uuid,
+        adminUserId,
```
I wanted it to be explicit so it's not just any userId (think of users that can be registering to someones website), bus specifically the admin users

fdel-car replied:
Going back to this, I still think the tracker needs to stay abstract, it doesn't care that it's an "admin" user.
For instance, if we use the same endpoint from the website for instance - which should happen - then sending `adminUserId` doesn't make sense there.

I'm just asking to change the tracker API, it's up to you if you still wanna call it `adminUserId` in the strapi code.

ivanThePleasant replied:
I can't agree that we need that abstraction so badly. We should only expect admin users on the tracker at any point, that's why I don't think renaming it makes much point. Still keen to explicitly show it should be only admin user and no other user. And it's fine to change it in the future should the need arise, but even in the website context we only send one event from it, which is meant to be anonymous.

fdel-car commented on packages/core/admin/admin/src/utils/unique-admin-hash.js:
```diff
@@ -0,0 +1,13 @@
+const hash = require('hash.js');
```
Not asking you to change but just wanted to note it somewhere, [Subtle Crypto](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto) is available in the browser, they do provide a `digest` function.

The only issue I can see with that is that it's only available in a secure context (HTTPS), which could be painful for our use case. Let's keep that in mind to reduce the bundle size.

ivanThePleasant replied:
Thanks for investigating that. I'll keep that in mind and we'll be able to swap it out if the solution works for our use case 😊👍

alexandrebodin commented:
Overall looks alright. 

Before moving forward we are introducing https://github.com/strapi/strapi/pull/14370.

This PR will getting the user directly in the sender in only one place in the code. that will make this PR a lot simpler and maintainable avoiding passing users everywhere.

ivanThePleasant commented:
Thanks @alexandrebodin, I will keep an eye out for when #14370 is merged 👍

fdel-car commented on packages/core/upload/server/controllers/admin-folder-file.js:
```diff
@@ -40,10 +40,13 @@ module.exports = {
-        totalFolderNumber,
-        totalAssetNumber: totalFileNumber + deletedFiles.length,
+        adminUser: ctx.state?.user,
+        groupProperties: {
```
I'm not confident those data should be in the `groupProperties` 🤔
@petersg83 I can see `deletedFiles` or `deletedFolders` which means that those data are associated with this specific event only right?

ivanThePleasant replied:
Even if it changes on an event, it will represent the state of how many folders/media is/are in the project at a given time, no? In my mind that falls under permanent group data

fdel-car replied:
No that's the thing the naming of those properties makes me say that they represent the number of deleted files and folders during the associated call.

fdel-car replied:
Just had a chat with him, and it's indeed what I thought it was. `rootFolderNumber` for instance is the number of folders selected when doing the bulk delete (thus deleted but only on the first level), and `totalFolderNumber` is the `rootFolderNumber` plus the number of folders that were nested inside the one the action deleted.

So as you can see this data is tightly linked to the event, it should be sent in the `eventProperties` putting it elsewhere will create confusion.

alexandrebodin commented:
> Thanks @alexandrebodin, I will keep an eye out for when #14370 is merged 👍

It would be nice if you can test using #14370 and see if we can make only one change in the sender() function like I think we can (for the server)

fdel-car commented on packages/core/admin/admin/src/pages/App/index.js:
```diff
@@ -89,18 +93,16 @@ function App() {
-                  environment: appInfo.currentEnvironment,
-                },
+                eventPropeties: {},
+                userProperties: { environment: appInfo.currentEnvironment },
```
The `environment` should be in the `eventProperties`, it's not tied to that user.

ivanThePleasant commented:
Yep, @alexandrebodin tested locally and can confirm it is possible to get the context in the sender as intended 👍

Convly commented on packages/core/strapi/lib/Strapi.js:
```diff
@@ -242,11 +242,14 @@ class Strapi {
+        database: strapi.config.get('database.connection.client'),
+        plugins: Object.keys(strapi.plugins),
+        // TODO: to add back
+        // providers: this.config.installedProviders,
```
What should we do with this one? Do we still want to send metrics on installed providers?

ivanThePleasant replied:
That I actually don't know, but it's not in the scope of the PR 😅

Convly commented on packages/core/strapi/lib/services/metrics/admin-user-hash.js:
```diff
@@ -0,0 +1,16 @@
+const hash = require('hash.js');
+
+const generateAdminUserHash = (ctx) => {
+  if (!ctx?.state?.user) return '';
```
Just a quick suggestion to stay consistent with the rest of the codebase

```suggestion
  if (!ctx?.state?.user) {
    return '';
  }
```

Convly commented on packages/core/strapi/lib/services/metrics/admin-user-hash.js:
```diff
@@ -0,0 +1,16 @@
+'use strict';
+
+const hash = require('hash.js');
```
Also, any reason why we're not using node's crypto standard package here?

```js
const crypto = require('crypto');

return crypto.createHash('sha256').update(ctx.state.user.email).digest('hex');
```

ivanThePleasant replied:
We use hash.js on admin code, so I wanted to be consistent

gu-stav replied:
@ivanThePleasant Just out of curiosity: where are we using this package? I can't find it anywhere ...

Convly commented on packages/core/strapi/package.json:
```diff
@@ -106,6 +106,7 @@
     "execa": "5.1.1",
     "fs-extra": "10.0.0",
     "glob": "7.2.0",
+    "hash.js": "1.1.7",
```
Why do we need this package?

Convly commented on packages/core/upload/server/controllers/admin-folder-file.js:
```diff
@@ -40,10 +40,12 @@ module.exports = {
-        rootAssetNumber: deletedFiles.length,
-        totalFolderNumber,
-        totalAssetNumber: totalFileNumber + deletedFiles.length,
+        groupProperties: {
```
I think that, based on @fdel-car comment, you've made the change from `groupProperties` to `eventProperties`, but only on **L234** and not on **L43**

gu-stav commented on packages/core/admin/admin/src/utils/unique-admin-hash.js:
```diff
@@ -0,0 +1,13 @@
+const hash = require('hash.js');
+
+const hashAdminUserEmail = (payload) => {
+  try {
+    const adminUserEmailHash = hash.sha256().update(payload.email).digest('hex');
+
+    return adminUserEmailHash;
+  } catch (error) {
+    return '';
+  }
+};
+
+module.exports = hashAdminUserEmail;
```
Modern browsers have a crypto API build it that we could utilize for that, instead of including quite a big dependency (5kB): https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest

What do you think about using this standard sha256 method, which then allows to backend to use a node API?

gu-stav replied:
The support for that is actually quite good. If you want to be super-secure you could check for the feature before using it, but I think with the support Matrix like that you won't run into problems ...

<img width="781" alt="Screenshot 2022-10-25 at 16 41 04" src="https://user-images.githubusercontent.com/2244375/197804541-f8dfa375-a281-4edc-8c46-b128cf6b0407.png">

ivanThePleasant replied:
Okay, sold 👍

gu-stav replied:
Great thanks. And sorry for being so late with the feedback - I just didn't know you were still working on it ...

ivanThePleasant replied:
Tracker had to be also configured to accept all these changes 👍

ivanThePleasant replied:
But I really appreciate the feedback, thank you 😊

ivanThePleasant replied:
@gu-stav have you implemented web crypto stuff previously? I'm looking at it and there are some bits I'm not sure about.
Firstly, the code looks a bit more complex. Conversions to a hex string is not looking pretty in particular.
Secondly, it only runs in HTTPS. But even from experience, when trying to set up Strapi in development on a VPS or something, misconfigs are normal and people end up running everything in HTTP for some time. This would mean we loose that bit of telemetry, not ideal.
What are your thoughts on this?

gu-stav replied:
I have not, but let me check. Regarding the secure context we won't run into problems on localhost:

> Locally-delivered resources such as those with http://127.0.0.1 URLs, http://localhost and http://*.localhost URLs (e.g. http://dev.whatever.localhost/), and file:// URLs are also considered to have been delivered securely.

https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts

I think I can't make a decision for deployments. Do we have data about that?

gu-stav replied:
This is the simplest solution I could come up with:

```js
const text = "Hello Ivan!";

function bufferToHex(buffer) {
  return [...new Uint8Array(buffer)]
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

async function digestMessage(message) {
  const msgUint8 = new TextEncoder().encode(message);
  const hashBuffer = await crypto.subtle.digest("SHA-256", msgUint8);
  return bufferToHex(hashBuffer);
}

(async () => {
  const hex = await digestMessage(text);
})();
```

I agree it looks a bit complex I agree, but on the other hand this can be isolated + tested well and if it helps to avoid adding ~6kB to the frontend-bundle I'd be willing to accept it.

ivanThePleasant replied:
It's deployments specifically. I don't think we have any data on that, but judging by my experience (especially when I had not a lot of it), having deployments set up can be a pain, especially when you need SSL, etc. So using it in non-secure context (especially if the use-case allows) is not unheard of.

gu-stav replied:
Alright, what do you think about this pattern in FE:

```js
function bufferToHex(buffer) {
  return [...new Uint8Array(buffer)]
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

async function digestMessage(message) {
  const msgUint8 = new TextEncoder().encode(message);
  const hashBuffer = await crypto.subtle.digest("SHA-256", msgUint8);
  return bufferToHex(hashBuffer);
}

const hash = require('hash.js');

const hashAdminUserEmail = async (payload) => {
  try {
    return await digestMessage(payload.email);
  } catch(error) {
    // not a secure context

    const hash = import('hash.js');
    return hash.sha256().update(payload.email).digest('hex');
  }
```

I'm not sure which type of error would be thrown in case of a non-secure context, but that's something we could find out I guess.

This would allow us to not increase the size of the main bundle and only load the dependency if it is needed.

ivanThePleasant replied:
We could try this, looks good! Thank you kindly 😊

fdel-car replied:
Cool outcome! We discussed this a bit [here](https://github.com/strapi/strapi/pull/14083#discussion_r962926271) also.
The thrown error will be a `TypeError` because `crypto.subtle` is gonna be `undefined`.

gu-stav commented on packages/core/admin/admin/src/pages/App/index.js:
```diff
@@ -89,18 +93,16 @@ function App() {
-            const deviceId = await getUID();
-
-            await fetch('https://analytics.strapi.io/track', {
+            fetch('https://analytics.strapi.io/api/v2/track', {
```
```suggestion
            await fetch('https://analytics.strapi.io/api/v2/track', {
```

Otherwise the rejected promise can not be fetched, which leads to an error will crash the whole UI.

ivanThePleasant replied:
Funnily enough, this is not my code and it has been working for all the time now 😅

gu-stav replied:
No it didn't :) I recently fixed all of these instances ...

ivanThePleasant replied:
But I merged master into this branch before pushing. I should have had all these changes if you fixed them already 🤔

gu-stav replied:
You should. Did you merge `master` or `main`? The change was there, so maybe it got lost in a merge conflict? Some of the other changes (the ones to `useTracking()` are there).

ivanThePleasant replied:
Main*. My bad. I've fixed all merge conflicts, had only one with await in the `useTracking()`. Didn't have any others

gu-stav commented on packages/core/strapi/lib/commands/opt-in-telemetry.js:
```diff
@@ -53,12 +53,12 @@ const generateNewPackageJSON = (packageObj) => {
 const sendEvent = async (uuid) => {
   try {
-    await fetch('https://analytics.strapi.io/track', {
+    await fetch('https://analytics.strapi.io/api/v2/track', {
```
Since this URL is used in multiple places: could we make it a constant?

ivanThePleasant replied:
Yes, this is a good point. Where should we put it though, since url is present across a lot of packages?

ivanThePleasant replied:
That's fine, but I'm not aware of where we keep any constants/configs of this sort in the monorepo. Could you point me towards it if you know? Or are you suggesting creating a space for it?

gu-stav commented on packages/core/admin/admin/src/utils/unique-admin-hash.js:
```diff
@@ -0,0 +1,21 @@
+  return bufferToHex(hashBuffer);
+}
+
+export default async function hashAdminUserEmail(payload) {
+  try {
+    return await digestMessage(payload.email);
+  } catch (error) {
+    // not a secure context
+    const hash = await import('hash.js');
+
+    return hash.sha256().update(payload.email).digest('hex');
+  }
+}
```
Do you think you could come up with a test for this utility? I could help you with that if you want, but I think it would be valuable to test, of the whole `catch` branch works.

ivanThePleasant replied:
Done

fdel-car commented on packages/generators/app/lib/utils/usage.js:
```diff
@@ -107,14 +118,12 @@ function trackError({ scope, error }) {
 }
 
 function trackUsage({ event, scope, error }) {
```
I think the `trackUsage` function should also benefit from the API change introduced to the telemetry sending function in the core. I'm talking about the possibility of also passing event & user properties.

The same could be said in the `useTracking` hook in the frontend.

ivanThePleasant replied:
Not sure what exactly you mean here? `trackUsage` has been changed to correspond to the API change. In `trackUsage` `getProperties` returns the correct new format of properties, but since in the generators all that info is static, there is no need to pass anything else to it. And in `useTracking` everything has been changed to correspond to the same way the backend sender does it. Am I missing something? 😅

fdel-car replied:
Both functions are now indeed compatible with the new format in the way they send the data.
But their input interface is not. The consumers of those functions do not have the possibility to categorize properties.

Again as I said, this requires too many changes most likely, so we can ignore that from now but let's keep it in mind. We would want to avoid discrepancies throughout the code on how the telemetry calls are made.

ivanThePleasant commented on packages/core/strapi/lib/services/metrics/sender.js:
```diff
@@ -42,41 +42,52 @@ module.exports = (strapi) => {
-          ...payload,
-          ...anonymousMetadata,
-        }),
+        eventProperties: adminUserId ? payload.eventProperties : {},
```
Should have only been on the userProperties, good spot

ivanThePleasant commented on packages/core/admin/admin/src/utils/unique-admin-hash.js:
```diff
@@ -0,0 +1,21 @@
+  return bufferToHex(hashBuffer);
+}
+
+export default async function hashAdminUserEmail(payload) {
```
`generateAdminUserId` sounds like something identifiable. It needs to convey from the name that it's generating something that is not identifiable. Since we're an open source product, I don't want to add something that could theoretically upset users

fdel-car replied:
I don't think users that are upset about such a function wouldn't give the said function a look.
It's up to you but I basically think we have 2 choices:
- Either you transform this function to make it completely abstract, like a simple SHA 256 hash function that could be reused where needed - pretty straightforward but the name then needs to show that.
- Or, what I've talked about above (and actually what you did), the retrieval of the admin user ID logic is in its own function. My only thought is that the name shouldn't technically recap the inner working of a function. You could consider this nitpicking maybe 🤷

ivanThePleasant commented on packages/core/admin/admin/src/components/AuthenticatedApp/index.js:
```diff
@@ -78,15 +78,17 @@ const AuthenticatedApp = () => {
 
   const shouldShowLoader = isLoading || shouldShowNotDependentQueriesLoader;
 
+  const adminUserId = userInfo ? hashAdminUserEmail(userInfo) : '';
```
I must have not changed it. In the early version I wasn't handling exceptions in the function, so had a ternary here. It's not needed anymore for sure, I'll change it

ivanThePleasant replied:
I think I forgot to delete this, since first version of the `hashAdminUserEmail` didn't handle absence of `userInfo`. It does now, so ternary not necessary

fdel-car commented on packages/core/admin/admin/src/utils/tests/uniqueAdminHash.test.js:
```diff
@@ -0,0 +1,55 @@
+  });
+
+  it('should return hash using crypto subtle', async () => {
+    Object.defineProperty(global.self, 'crypto', {
```
I'm not sure I get the point of having added the @peculiar/webcrypto package in our case. You only use it here for our tests, meaning that what you're doing is testing the validity of that package which we are not actually using.

Instead, why not just mock - once in that case - the `global.window.crypto` object? You don't even have to truly mock the implementation, the only important thing here is that the correct function gets called and that it returns a value of the correct type (an `ArrayBuffer` for the `digest` function in our case).

ivanThePleasant replied:
Removed it

alexandrebodin commented on packages/core/helper-plugin/lib/src/hooks/useTracking/index.js:
```diff
@@ -5,22 +5,31 @@ import useAppInfos from '../useAppInfos';
+          'https://analytics.strapi.io/api/v2/track',
+          {
+            event,
+            adminUserId,
```
Should this just be called userId ?

ivanThePleasant replied:
@fdel-car was also advocating for this. Changed it 👍

alexandrebodin commented on packages/core/strapi/lib/services/metrics/admin-user-hash.js:
```diff
@@ -0,0 +1,14 @@
+const crypto = require('crypto');
+
+const generateAdminUserHash = (ctx) => {
+  if (!ctx?.state?.user) {
```
Why not use the requestContext here to get the user and avoid having to pass it as a param ?

ivanThePleasant replied:
Changed it 👍 Although this would leave admin implementation of this still passing a param

alexandrebodin commented on packages/core/helper-plugin/lib/src/hooks/useTracking/index.js:
```diff
@@ -5,22 +5,31 @@ import useAppInfos from '../useAppInfos';
+            eventProperties: { ...properties },
+            userProperties: {},
+            groupProperties: {
+              ...telemetryProperties,
```
Are we sure those are only group Properties? it looks lik eit could contain event/user & group properties eventually

fdel-car replied:
We discussed it [here](https://github.com/strapi/strapi/pull/14083#discussion_r1006919330) and I agree, but it's likely that it will involve many changes.

alexandrebodin replied:
the telemetry endpoint could simply return an object containing 

groupProp and here we would use a merge instead of a spread. Not sure why this would require more changes? what do you have in mind ?

fdel-car replied:
Oh, I see, you're only talking about what's coming from the context, and indeed yeah the changes should be kept minimal.

I was talking about allowing the user to send all kinds of properties when using the `useTracking` hook, cause right now we can only pass event properties.

alexandrebodin replied:
Yes not the same thing but this could be interesting too allow indeed. 

We can refactor that code when we will need to send other properties anyway

ivanThePleasant replied:
I'll resolve this conversation then since as I understand no actions are currently necessary. For this time those are indeed the only `groupProperties`, but that can be refactored in the future should there be a need 😊👍

fdel-car replied:
Unless I misunderstood, I think he wanted to have a different format for the `telemetryProperties` coming from the context so that not everything simply ends up in group properties.

alexandrebodin reviewed (commented):
A few questions left but overall looks good

joshuaellis commented on packages/core/admin/admin/src/components/AuthenticatedApp/index.js:
```diff
@@ -78,15 +78,17 @@ const AuthenticatedApp = () => {
   const shouldShowLoader = isLoading || shouldShowNotDependentQueriesLoader;
 
+  const adminUserId = hashAdminUserEmail(userInfo);
   const appInfosValue = useMemo(() => {
```
Can we remove the use of memo? I don't personally see the point in using it to construct an object with no transformations?

ivanThePleasant replied:
Is there any harm in keeping it in your opinion? I don't personally know why it is used here, but this lack of context could delay this PR even further in case it needs to be reverted back, that's my only concern.

joshuaellis replied:
This optimisation isn't free and in this circumstance I believe it's probably doing more harm than good. In the FE sync and off the back it was agreed we would remove these premature optimisations.

> I don't personally know why it is used here

No one does, that's the other problem.

ivanThePleasant replied:
Gotcha, thanks. I wasn't at the FE sync, so didn't know that was agreed upon. In that case I can refactor it and give it a try

joshuaellis replied:
Thanks @ivanThePleasant 😄

ivanThePleasant replied:
Hey @joshuaellis annoyingly we have some eslint rules that will scream at me if I want to get rid of this memo, and that will need to be fixed via our eslint package I guess. Because this is not part of the scope of this PR, I'll leave it as is for now, but will remove it in a separate PR

alexandrebodin commented on packages/core/admin/admin/src/utils/uniqueAdminHash.js:
```diff
@@ -0,0 +1,24 @@
+    return '';
+  }
+  try {
+    return await utils.digestMessage(payload.email);
```
In what cases do we need hash.js ?

gu-stav replied:
You can follow the discussions above: the browser crypto APIs only work in secure contexts (localhost/ https) and would fail if someone deploys Strapi without using https. If there is a way to get rid of the code-branch I'd be very happy, but I think Ivan had a valid point ...

alexandrebodin replied:
I'm not sure Strapi even works in production deployments if it's not https 🤔  did someone confirmed this ? @ivanThePleasant ?

alexandrebodin replied:
What do you mean by development ? localhost is considered a secure context and will be able to use the crypto already. 

I wasn't asking about do people do it or not but rather can they even do it ? I'm not sure Strapi will work once deployed on a server if it doesn't have https based on our security middleware. Did you try a simple deploy on a VPS without ssl ?

ivanThePleasant replied:
I'm pretty sure I've done something like this before on a VPS (not localhost), let me double check and come back on this 👍 Maybe I am confusing something

alexandrebodin replied:
Thanks 🙏

ivanThePleasant replied:
Hey @alexandrebodin yes, I've checked it and you're correct, admin panel is not accessible without secure context, so it has to be served over https. So we can get rid of the package then. I don't think there are a lot of people who would try to do something to deliberately serve admin panel over http somehow. And apologies for the delay, I found another bug with this, had to fix it as well

gu-stav replied:
Great to hear. That means we can also get rid of the package for the backend part, don't we?

ivanThePleasant replied:
I already got rid of it, we use node crypto there now

alexandrebodin reviewed (approved):
This looks good. I need the PM approval on this before we merge

ivanThePleasant commented:
@alexandrebodin I have removed the package and also changed admin `userId` hashing to be awaited, but apart from that no other changes were made, so if you could approve once again that would be great. As to PM approval, Nkem is working on that.

@gu-stav as we discussed regarding the `useEffect` for awaiting admin user hashing, if you could also quickly have a look and approve, that would be great 😊👍

alexandrebodin commented on packages/core/admin/admin/src/components/AuthenticatedApp/index.js:
```diff
@@ -31,6 +31,7 @@ const AuthenticatedApp = () => {
   const userInfo = auth.getUserInfo();
   const userName = get(userInfo, 'username') || getFullName(userInfo.firstname, userInfo.lastname);
   const [userDisplayName, setUserDisplayName] = useState(userName);
+  const [userId, setUserId] = useState('');
```
use not use null as default value instead of an empty string ?

ivanThePleasant replied:
Used empty string because we have an empty string everywhere, where events are anonymous (or malfunctioning for some reason), so just kept consistency. Another small reason is that it indicates it should be a string. But it can be null if that's better for code consistency across the code base

alexandrebodin replied:
Yeah overall if there is no value you would rather want to use a null|undefined to represent "no-value" rather than an empty string. Do you need to convert it to a string when sending the events ?

ivanThePleasant replied:
Nope, Amplitude will be fine with null. I will change it 👍

alexandrebodin commented on packages/core/admin/admin/src/utils/uniqueAdminHash.js:
```diff
@@ -17,8 +17,6 @@ export default async function hashAdminUserEmail(payload) {
   try {
     return await utils.digestMessage(payload.email);
   } catch (error) {
-    const hash = await import('hash.js');
```
you removed the code but I don't see the package being removed from the dependencies ?

gu-stav commented on packages/core/admin/admin/src/utils/uniqueAdminHash.js:
```diff
@@ -17,8 +17,6 @@ export default async function hashAdminUserEmail(payload) {
-    const hash = await import('hash.js');
-
-    return hash.sha256().update(payload.email).digest('hex');
+    return '';
```
Do you handle this case somewhere else? I'd be afraid we end up with a few empty values in Amplitude without knowing why.

ivanThePleasant replied:
Not really, but I also don't see a viable way to handle this other way. Don't know why it would, but in case subtle crypto fails, there is no other way to hash user email on the admin.  I could add an obvious string, something like `anonymousAdminUser` so we could quickly see on Amplitude in case this specific thing is failing. But on the other hand if we start seeing non-anonymous admin events without a user id on Amplitude, this would be the only place to look for.

gu-stav replied:
Could we not send the tracking event, if the hash can not be generated? I don't know if that makes sense and leave it up to you :) I was just wondering if it would be good to handle this case for the sake of data-consistency.

ivanThePleasant replied:
I see your point, but I also need to add some context for reassurance. It's better to send the event regardless in my opinion, because even without the user id available (in case that happens), there's going to be a `deviceId` available on the data received by amplitude. That would allow it to map the events to users retrospectively if we need to push some sort of fix for the matter. But I don't expect us needing to do it in the first place

gu-stav commented on packages/core/admin/package.json:
```diff
@@ -77,6 +77,7 @@
     "fork-ts-checker-webpack-plugin": "7.2.1",
     "formik": "^2.2.6",
     "fs-extra": "10.0.0",
+    "hash.js": "1.1.7",
```
As Alex pointed out, this can be removed.

gu-stav reviewed (approved):
Great work! Sorry for all the back and forth, and thanks for keeping up the good work!

alexandrebodin commented on packages/core/admin/admin/src/utils/uniqueAdminHash.js:
```diff
@@ -0,0 +1,22 @@
+  try {
+    return await utils.digestMessage(payload.email);
+  } catch (error) {
+    return '';
```
shouldn't those return null too then ?

ivanThePleasant replied:
Changed

alexandrebodin commented on packages/core/admin/admin/src/utils/uniqueAdminHash.js:
```diff
@@ -0,0 +1,22 @@
+
+export default async function hashAdminUserEmail(payload) {
+  if (!payload) {
+    return '';
```
shouldn't those return null too then ?

ivanThePleasant replied:
Changed

alexandrebodin reviewed (approved):
LGTM 👍

ivanThePleasant commented:
@alexandrebodin thanks! Is it ok now to remove don't merge flag?
